### DISCUSSION TITLE
Understanding uses of NavigationService and why it won't work from a NotificationDelegate

### DISCUSSION THREAD

**Discussion by munkii** [PrismLibrary/Prism#3211] [.NET MAUI] [answered]
I am still in the process of migrating from FreshMVVM and Xamarin Forms and defiently still don't fully understand Navigation, Prism and MAUI.

Our app responds to 4 or 5 different "types" of PushNotification. The logic for how to deal with those is triggered from the Shiny PushDelegate. Each Action has a number of dependencies and the one I am currently trying to debug needs an INavigationService

```
public class SendUserToSignalViewModelNotificationAction : INotificationAction
{
    private readonly INavigationService navigationService;

    public string ActionName => "sendusertosignalviewmodel";

    /// <summary>
    /// Initializes a new instance of the <see cref="SendUserToSignalViewModelNotificationAction"/> class.
    /// </summary>
    public SendUserToSignalViewModelNotificationAction(INavigationService navigationService)
    {
        this.navigationService = navigationService;
    }

    public bool CallableAtStage(NotificationStage stage)
    {
        return stage == NotificationStage.OnEntry;
    }

    /// <inheritdoc/>
    public async Task TriggerAsync(string notificationTitle, string notificationMessage, string[] data)
    {
        var signalState = GetSignalStateFromActionData(data);

        Type viewModelType = Helpers.SignalStateToViewModelHelper.GetSignalDetailViewModelType((State)Enum.Parse(typeof(State), signalState));
        await this.navigationService.NavigateAsync(viewModelType, null, false, true);
    }

    private static string GetSignalStateFromActionData(string[] data)
    {
       //  removed for brevity
    }

    private static string GetKeyedDataFromActionData(string key, string[] data)
    {
        //removed for brevity
    }
}
```

The call to `navigationService.NavigateAsync` does result in the user being navigated to the expected View/ViewModel however it appears like it is in a new NavigationStack of just the one page, despite the app already being loaded and having an existing stack.

If I switch away from trying to use the injected INavigationService and instead publish an event that I subscribe to in an already loaded ViewModel the navigation works exactly as I expect and the stack is maintained.

What is the difference between using the INavigationService that is getting injected into my `SendUserToSignalViewModelNotificationAction ` and the one that was injected into the previously loaded ViewModel. I appreciate this is probably a laughably basic question.

**Comment by dansiegel** [comment, marked as answer]
The NavigationService is not a Singleton Service. It is a scoped service that is created for EACH Page in the application. I believe @aritchie has a solution for this... perhaps he can elaborate on the way to handle this in Shiny.

**Reply by aritchie** [reply]
I have a reasonable solution that works in standard cases - https://github.com/shinyorg/mediator/blob/main/src/Shiny.Mediator.Prism/IGlobalNavigationService.cs

The notification delegate entry can fire BEFORE android is ready to navigate or render anything.  There are ways to await an activity being ready, but that's a topic for another day.
